### ISSUE TITLE
Make cuDNN/cuTENSOR/NCCL optional

### ISSUE THREAD

**Issue by pentschev** [conda-forge/cupy-feedstock#109]
Currently CuPy lists cuDNN as a requirement, but given cuDNN is large in size it's possibly a good idea to make it optional instead. A concrete use case is RAPIDS, it depends on CuPy but doesn't use cuDNN, which forces the inclusion of that library in official Docker containers, etc.

@leofang mentioned there may be a few things that need to be changed. Could we discuss those here? Do we need those changes in the CuPy repo or do they need to be here in the feedstock?

cc @kkraus14 @kmaehashi @jakirkham

**Comment by jakirkham** [comment]
Wee need changes in both ( https://github.com/cupy/cupy/issues/4850#issuecomment-794426626 ). The CuPy bits can be discussed in that issue. The Conda bits are easy to handle once the CuPy bit are done (and in a release)

**Comment by jakirkham** [comment]
Does CuPy link to cuTENSOR at build time?

**Comment by leofang** [comment]
Yes, there're two Cython modules for supporting cuTENSOR.

**Comment by leofang** [comment]
> Yes, there're two Cython modules for supporting cuTENSOR.

Likewise for cuDNN and NCCL.

**Comment by jakirkham** [comment]
What happens if a user tries to load those modules if they were linked to those libraries, but the libraries were not present? A library load error? Or does CuPy have some way to insulate the user from that?

**Comment by jakirkham** [comment]
Sounds good. I think if we have that in place for cuTENSOR and NCCL, it will be easy to make those optional here in the same way

**Comment by jakirkham** [comment]
Also ok with dropping CentOS 6 (assuming I understood your suggestion above correctly)

**Comment by jakirkham** [comment]
Once that is done, I think we can add all of these to [`ignore_run_exports`]( https://docs.conda.io/projects/conda-build/en/latest/resources/define-metadata.html#export-runtime-requirements ) and [`run_constrained`]( https://docs.conda.io/projects/conda-build/en/latest/resources/define-metadata.html#run-constrained ) that way we still build against them, but not ship them by default. Users can then optionally install these extra pieces as they need them

**Comment by leofang** [comment]
In case it wasn't clear, the work plan is as follows:
1. Get the PR https://github.com/cupy/cupy/pull/4873 in so that we can give instructions to CF users at runtime for installing missing libraries.
2. Generate a json file in the recipe here (likely when the recipe is updated next time, say for v8.6.0/v9.0.0rc1) to enable the warning mechanism. Also use the tricks John mentioned above to do the actual opt-out job.

**Comment by leofang** [comment]
(btw @pentschev I changed the issue title to reflect the new plan.)

**Comment by leofang** [comment]
@pentschev this is done starting v9.0.0rc1; v8.6 still has to carry the deps...

**Comment by pentschev** [comment]
Thanks so much @leofang , you're awesome! 😄
